### PR TITLE
MTDSA-7646: Updated HATEOAS for benefits with duplicate benefit id

### PR DESCRIPTION
**Pull request by narendravijayarao**

### CODE REVIEW

File: app/v1/controllers/ListBenefitsController.scala [hunk 1]
```diff
@@ -103,4 +103,19 @@ class ListBenefitsController @Inject()(val authService: EnrolmentsAuthService,
       case DownstreamError => InternalServerError(Json.toJson(errorWrapper))
     }
   }
+
+  private def process(response: ListBenefitsResponse[StateBenefit], benefitId: Option[String]): ListBenefitsResponse[StateBenefit] = {
```
Comment (MarkAKelly) [inline app/v1/controllers/ListBenefitsController.scala:107]: Ambiguous function name. Lets call it something very explicit, as its a very explicit operation.

File: app/v1/controllers/ListBenefitsController.scala [hunk 2]
```diff
@@ -103,4 +103,19 @@ class ListBenefitsController @Inject()(val authService: EnrolmentsAuthService,
...
+      case None => response
+      case _ => (response.stateBenefits, response.customerAddedStateBenefits) match {
+        case (Some(hmrc), Some(custom)) if hasCommon(hmrc, custom) != Nil =>
+          response.copy(Some(hmrc.map(_.copy(isCommon = true))), Some(custom.map(_.copy(isCommon = true))))
+        case (_, _) => response
+      }
+    }
+  }
+
+  private def hasCommon(hmrcB: Seq[StateBenefit], customB: Seq[StateBenefit]): List[StateBenefit] =
```
Comment (MarkAKelly) [inline app/v1/controllers/ListBenefitsController.scala:119]: also this function doesnt return a boolean, which you would expect it to by its name.  If what we are doing here is retrieving common benefits between the two lets call it something like 'listCommonBenefits'

File: app/v1/controllers/ListBenefitsController.scala [hunk 3]
```diff
@@ -69,7 +69,7 @@ class ListBenefitsController @Inject()(val authService: EnrolmentsAuthService,
           hateoasResponse <- EitherT.fromEither[Future](
             hateoasFactory
               .wrapList(
-                serviceResponse.responseData,
+                process(serviceResponse.responseData, benefitId),
```
Comment (MarkAKelly) [inline app/v1/controllers/ListBenefitsController.scala:72]: Should we even be doing this at the controller layer?  As the logic relates to the models, is it not better that this sits inside the ListBenefitsResponse class with the actual link logic?

Comment (narendravijayarao) [reply app/v1/controllers/ListBenefitsController.scala:72]: Nope, we are applying HATEOAS to the response in the controller and it depends on the GET/LIST call. We can move the code to model, but ...

Comment (MarkAKelly) [reply app/v1/controllers/ListBenefitsController.scala]: but?

Comment (narendravijayarao) [reply app/v1/controllers/ListBenefitsController.scala]: modifying the model inside a function of the model object is really awful :)

**Pull request merged**